File: cassandra/python/cassandra_example.py
# Copyright (c) 2018 Aiven, Helsinki, Finland. https://aiven.io/

import ssl

from cassandra.auth import PlainTextAuthProvider
from cassandra.cluster import Cluster
from cassandra.policies import DCAwareRoundRobinPolicy


def cassandra_example(args):
    auth_provider = PlainTextAuthProvider(args.username, args.password)
    ssl_options = {"ca_certs": args.ca_path, "cert_reqs": ssl.CERT_REQUIRED}
    with Cluster([args.host], port=args.port, ssl_options=ssl_options, auth_provider=auth_provider,
                 load_balancing_policy=DCAwareRoundRobinPolicy(local_dc='aiven')) as cluster:
        with cluster.connect() as session:
            # Create a keyspace
            session.execute("""
                CREATE KEYSPACE IF NOT EXISTS example_keyspace
                WITH REPLICATION = {'class': 'NetworkTopologyStrategy', 'aiven': 3}
            """)

            # Create a table
            session.execute("""
                CREATE TABLE IF NOT EXISTS example_keyspace.example_table (
                    id int PRIMARY KEY,
                    message text
                )
            """)

            # Insert some data
            for i in range(10):
                session.execute("""
                    INSERT INTO example_keyspace.example_table (id, message)
                    VALUES (%s, %s)
                """, (i, "Hello world!"))

            # Read it back
            for row in session.execute("SELECT id, message FROM example_keyspace.example_table"):
                print("Row: id = {}, message = {}".format(row.id, row.message))

File: cassandra/go/cassandra_example.go
// Copyright (c) 2018 Aiven, Helsinki, Finland. https://aiven.io/

package main

import (
	"fmt"
	"github.com/gocql/gocql"
	"log"
	"time"
)

func cassandraExample(args Args) {
	var session = createSession(args)
	defer session.Close()
	createSchema(session)
	writeData(session)
	readData(session)
}

func createSession(args Args) gocql.Session {
	cluster := gocql.NewCluster(args.Host)
	cluster.ConnectTimeout = 10 * time.Second
	cluster.Port = args.Port
	cluster.ProtoVersion = 4
	cluster.Authenticator = gocql.PasswordAuthenticator{
		Username: args.Username,
		Password: args.Password,
	}
	cluster.SslOpts = &gocql.SslOptions{
		CaPath: args.CaPath,
	}
	cluster.Consistency = gocql.Quorum

	session, err := cluster.CreateSession()
	if err != nil {
		log.Fatal(err)
	}
	return *session
}

func createSchema(session gocql.Session) {
	if err := session.Query(
		"CREATE KEYSPACE IF NOT EXISTS example_keyspace WITH REPLICATION = {'class': 'NetworkTopologyStrategy', 'aiven': 3}",
	).Exec(); err != nil {
		log.Fatal(err)
	}
	if err := session.Query(
		"CREATE TABLE IF NOT EXISTS example_keyspace.example_go (id int PRIMARY KEY, message text)",
	).Exec(); err != nil {
		log.Fatal(err)
	}
}

func writeData(session gocql.Session) {
	for i := 1; i <= 10; i++ {
		if err := session.Query(
			"INSERT INTO example_keyspace.example_go (id, message) VALUES (?, ?)", i, "Hello from golang!",
		).Exec(); err != nil {
			log.Fatal(err)
		}
	}
}

func readData(session gocql.Session) {
	iter := session.Query("SELECT id, message FROM example_keyspace.example_go").Iter()
	var id int
	var message string
	for iter.Scan(&id, &message) {
		fmt.Printf("Row: id = %d, message = %s\n", id, message)
	}
	if err := iter.Close(); err != nil {
		log.Fatal(err)
	}
}
